e2e/pipelines/test_eval_rag_pipelines.py

```diff
@@ -0,0 +1,145 @@
+    {% for doc in documents %}
+        {{ doc.content }}
+    {% endfor %}
+
+    \nQuestion: {{question}}
+    \nAnswer:
+    """
+    rag_pipeline = Pipeline()
+    rag_pipeline.add_component(
+        instance=SentenceTransformersTextEmbedder(model_name_or_path="sentence-transformers/all-MiniLM-L6-v2"),
+        name="text_embedder",
+    )
+    rag_pipeline.add_component(
+        instance=InMemoryEmbeddingRetriever(document_store=InMemoryDocumentStore()), name="retriever"
+    )
+    rag_pipeline.add_component(instance=PromptBuilder(template=prompt_template), name="prompt_builder")
+    rag_pipeline.add_component(
+        instance=HuggingFaceLocalGenerator(
+            model_name_or_path="google/flan-t5-small",
+            task="text2text-generation",
+            generation_kwargs={"max_new_tokens": 100, "temperature": 0.5, "do_sample": True},
+        ),
+        name="llm",
+    )
+    rag_pipeline.add_component(instance=AnswerBuilder(), name="answer_builder")
+    rag_pipeline.connect("text_embedder", "retriever")
+    rag_pipeline.connect("retriever", "prompt_builder.documents")
+    rag_pipeline.connect("prompt_builder", "llm")
+    rag_pipeline.connect("llm.replies", "answer_builder.replies")
+    rag_pipeline.connect("retriever", "answer_builder.documents")
+
+    # Populate the document store
+    documents = [
+        Document(content="My name is Jean and I live in Paris."),
+        Document(content="My name is Mark and I live in Berlin."),
+        Document(content="My name is Giorgio and I live in Rome."),
+    ]
+    document_store = rag_pipeline.get_component("retriever").document_store
+    indexing_pipeline = Pipeline()
+    indexing_pipeline.add_component(
+        instance=SentenceTransformersDocumentEmbedder(model_name_or_path="sentence-transformers/all-MiniLM-L6-v2"),
+        name="document_embedder",
+    )
+    indexing_pipeline.add_component(instance=DocumentWriter(document_store=document_store), name="document_writer")
+    indexing_pipeline.connect("document_embedder", "document_writer")
+    indexing_pipeline.run({"document_embedder": {"documents": documents}})
+
+    # Query and assert
+    questions = ["Who lives in Paris?", "Who lives in Berlin?", "Who lives in Rome?"]
+    inputs = [
+        {
+            "prompt_builder": {"question": question},
+            "text_embedder": {"text": question},
+            "answer_builder": {"query": question},
+        }
+        for question in questions
+    ]
+
+    expected_outputs = [
+        {"llm": {"replies": ["Jean"]}},
+        {"llm": {"replies": ["Mark"]}},
+        {"llm": {"replies": ["Giorgio"]}},
+    ]
+
+    eval_result = eval(rag_pipeline, inputs=inputs, expected_outputs=expected_outputs)
+
+    assert eval_result.inputs == inputs
+    assert eval_result.expected_outputs == expected_outputs
+    assert len(eval_result.outputs) == len(expected_outputs) == len(inputs)
+    assert eval_result.runnable == rag_pipeline
```

Comment:
How should I compare two Pipelines/Components, is comparing Pipelines using `==` not allowed?

When comparing the pipeline instances, I get the following error:

```bash
>       assert eval_result.runnable == rag_pipeline

e2e/pipelines/test_eval_rag_pipelines.py:145: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
haystack/core/pipeline/pipeline.py:87: in __eq__
    and self._comparable_nodes_list(self.graph) == self._comparable_nodes_list(other.graph)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <haystack.pipeline.Pipeline object at 0x7f2a44acd310>
graph = <networkx.classes.multidigraph.MultiDiGraph object at 0x7f2a44acd580>

    def _comparable_nodes_list(self, graph: networkx.MultiDiGraph) -> List[Dict[str, Any]]:
        """
        Replaces instances of nodes with their class name in order to make sure they're comparable.
        """
        nodes = []
        for node in graph.nodes:
            comparable_node = graph.nodes[node]
            comparable_node["instance"] = comparable_node["instance"].__class__
            nodes.append(comparable_node)
>       nodes.sort()
E       TypeError: '<' not supported between instances of 'dict' and 'dict'
```

Reply:
Probably the best thing would be converting both to `dict` using `to_dict()`.
Something like this:
```
assert eval_result.runnable.to_dict() == rag_pipeline.to_dict()
```

Reply:
Thanks! That looks like the least error-prone as well. I'll update the tests with the changes.